virt-controller/watch/vmi: propagate vmi.Spec.TerminationGracePeriodSeconds to launcher DeleteOptions

dasionov opened the pull request:
### What this PR does
When deleting the launcher, we should use `vmi.Spec.TerminationGracePeriodSeconds` in the DeleteOptions.
without it the launcher pod  will have the grace preiod from the initial manifest.
This limits users who start  with a high grace period in a template but later want to terminate quickly.
With a simple patch of the  `TerminationGracePeriodSeconds`  field, users will be able to override the grace time.

#### Before this PR:
- The `TerminationGracePeriodSeconds` value from the VMI spec was not  used in `DeleteOptions` when deleting the `virt-launcher` pod. 
- The launcher pod always used the value defined in its initial manifest.  
- Users could not change the termination grace period dynamically. 

#### After this PR:
- The `TerminationGracePeriodSeconds` field is now propagated to `DeleteOptions` when deleting the `virt-launcher` pod.  
- Users can patch the `TerminationGracePeriodSeconds` field in the VM and have it affect the pod's graceful termination time.
- This allows quicker termination of VMs when needed.

In the future it will be possible to create a virctl command like:

`virctl delete myvm --grace-preiod=0` which will patch the grace time from the flag and issue a delete api call.


### Release note
```release-note
None
```

ksimon1 commented on pkg/virt-controller/watch/vmi/vmi.go:
```diff
@@ -1524,8 +1524,18 @@ func (c *Controller) deleteAllMatchingPods(vmi *virtv1.VirtualMachineInstance) e
 			continue
 		}
 
+		var gracePeriod *int64
```
Assign zero directly to gracePeriod and change it only when vmi.Spec.TerminationGracePeriodSeconds != nil?

dasionov replied:
Thanks, fixed.

ksimon1 commented on tests/vm_test.go:
```diff
@@ -821,6 +821,66 @@ var _ = Describe("[rfe_id:1177][crit:medium][vendor:cnv-qe@redhat.com][level:com
+					}, 30*time.Second, 5*time.Second).Should(Succeed(), "virt-launcher pod should be deleted after grace period expires")
+				},
+					Entry("with immediate deletion (grace period 0)", int64(0)),
+					Entry("with a 20s grace period", int64(20)),
```
would it be possible to decrease the value to e.g. 5 seconds?

dasionov replied:
done

Barakmor1 commented on tests/vm_test.go:
```diff
@@ -821,6 +821,66 @@ var _ = Describe("[rfe_id:1177][crit:medium][vendor:cnv-qe@redhat.com][level:com
 			})
+
+			When("Deleting a VM with high TerminationGracePeriod", func() {
+				var vm *v1.VirtualMachine
+				var pod *k8sv1.Pod
+				BeforeEach(func() {
+					By("Creating a VM with a high TerminationGracePeriod")
+					vm = libvmi.NewVirtualMachine(
+						libvmifact.NewGuestless(libvmi.WithTerminationGracePeriod(1600)),
+						libvmi.WithRunStrategy(v1.RunStrategyAlways))
+					vm.Namespace = testsuite.GetTestNamespace(vm)
+
+					vm, err = virtClient.VirtualMachine(vm.Namespace).Create(context.Background(), vm, metav1.CreateOptions{})
+					Expect(err).ToNot(HaveOccurred())
+
+					By("Waiting for VMI to start")
+					Eventually(ThisVM(vm), 240*time.Second, 1*time.Second).Should(BeReady())
+
+					vmi, err := virtClient.VirtualMachineInstance(vm.Namespace).Get(context.Background(), vm.Name, metav1.GetOptions{})
+					Expect(err).ToNot(HaveOccurred())
+
+					pod, err = libpod.GetPodByVirtualMachineInstance(vmi, vmi.Namespace)
+					Expect(err).ToNot(HaveOccurred())
+				})
```
Nit: I think we can do this in the test itself instead of in a separate beforeEach scope.

dasionov replied:
ok

Barakmor1 commented on tests/vm_test.go:
```diff
@@ -821,6 +821,66 @@ var _ = Describe("[rfe_id:1177][crit:medium][vendor:cnv-qe@redhat.com][level:com
+						g.Expect(*vmi.Spec.TerminationGracePeriodSeconds).To(BeEquivalentTo(gracePeriod))
+					}, 60*time.Second, 5*time.Second).Should(Succeed(), "VMI should have been updated by the patch")
+
+					By("Deleting the VM")
+					err = virtClient.VirtualMachine(vm.Namespace).Delete(context.Background(), vm.Name, metav1.DeleteOptions{})
+					Expect(err).ToNot(HaveOccurred())
+
+					if gracePeriod > 0 {
+						By("Ensuring the pod is still present before the grace period expires")
+						Consistently(func(g Gomega) {
+							_, err := virtClient.CoreV1().Pods(vm.Namespace).Get(context.Background(), pod.Name, metav1.GetOptions{})
+							g.Expect(err).ToNot(HaveOccurred())
+						}, time.Duration(gracePeriod)*time.Second, 5*time.Second).Should(Succeed(), "virt-launcher pod should not be deleted before grace period expires")
+					}
+
+					Eventually(func(g Gomega) {
+						_, err := virtClient.CoreV1().Pods(vm.Namespace).Get(context.Background(), pod.Name, metav1.GetOptions{})
+						g.Expect(errors.IsNotFound(err)).To(BeTrue())
+					}, 30*time.Second, 5*time.Second).Should(Succeed(), "virt-launcher pod should be deleted after grace period expires")
+				},
+					Entry("with immediate deletion (grace period 0)", int64(0)),
+					Entry("with a 5s grace period", int64(5)),
+				)
```
I'm a bit worried this might be flaky. Do you think we can reduce this to a single functional test that patches the grace period to 0 and verifies that the pod eventually gets deleted in less than 5 seconds? To me, this test case is enough to verify it

dasionov replied:
yea

Barakmor1 commented on tests/vm_test.go:
```diff
@@ -821,6 +821,66 @@ var _ = Describe("[rfe_id:1177][crit:medium][vendor:cnv-qe@redhat.com][level:com
+					Expect(err).ToNot(HaveOccurred())
+
+					By("Ensuring the VMI reflects the updated TerminationGracePeriodSeconds")
+					Eventually(func(g Gomega) {
+						vmi, err := virtClient.VirtualMachineInstance(vm.Namespace).Get(context.Background(), vm.Name, metav1.GetOptions{})
+						g.Expect(err).ToNot(HaveOccurred())
+						g.Expect(*vmi.Spec.TerminationGracePeriodSeconds).To(BeEquivalentTo(gracePeriod))
+					}, 60*time.Second, 5*time.Second).Should(Succeed(), "VMI should have been updated by the patch")
```
WDYT about checking the launcher pod instead?

dasionov replied:
It might be too flaky because the launcher pod is updated only after the delete API call. If I check the grace period value beforehand, it could be a race condition with the deletion itself—the pod might already be gone. This means the test could still be working as expected, but I want to ensure the VMI holds the correct value before proceeding with deletion.

Barakmor1 commented on pkg/virt-controller/watch/vmi/vmi.go:
```diff
@@ -1524,8 +1525,15 @@ func (c *Controller) deleteAllMatchingPods(vmi *virtv1.VirtualMachineInstance) e
 		}
 
+		gracePeriod := pointer.P(int64(0))
+		if strings.HasPrefix(pod.Name, "virt-launcher") && vmi.Spec.TerminationGracePeriodSeconds != nil {
+			gracePeriod = vmi.Spec.TerminationGracePeriodSeconds
+		}
+
 		c.podExpectations.ExpectDeletions(vmiKey, []string{controller.PodKey(pod)})
-		err := c.clientset.CoreV1().Pods(vmi.Namespace).Delete(context.Background(), pod.Name, v1.DeleteOptions{})
+		err := c.clientset.CoreV1().Pods(vmi.Namespace).Delete(context.Background(), pod.Name, v1.DeleteOptions{
+			GracePeriodSeconds: gracePeriod,
+		})
```
I'm wondering about the attached pods. Wouldn't it be reasonable to set the TerminationGracePeriodSeconds for any pod associated with the VM?

dasionov replied:
@xpivarc Can you think of a scenario where the launcher pod is attempting to terminate gracefully, but it requires the attachment pod to remain running for as long as the launcher is still alive?

Barakmor1 reviewed (commented):
Thanks for addressing this bug! I’ve left a few comments.

dasionov commented:
/test pull-kubevirt-check-tests-for-flakes
/test pull-kubevirt-e2e-k8s-1.32-sig-compute
/test pull-kubevirt-e2e-k8s-1.32-sig-storage

dasionov commented:
/test pull-kubevirt-e2e-k8s-1.31-sig-compute-migrations

dasionov commented:
/test pull-kubevirt-e2e-k8s-1.31-sig-compute-migrations

alicefr commented on pkg/virt-controller/watch/vm/vm_test.go:
```diff
@@ -6127,6 +6127,103 @@ var _ = Describe("VirtualMachine", func() {
+					patchAction, ok := action.(testing.PatchAction)
+					Expect(ok).To(BeTrue())
+
+					expectedPatch := []byte(`[{"op":"replace","path":"/spec/terminationGracePeriodSeconds","value":60}]`)
```
Please don't use hardcoded patches, but instead rely on the patchset package

dasionov replied:
done

alicefr commented on pkg/virt-controller/watch/vm/vm_test.go:
```diff
@@ -6127,6 +6127,103 @@ var _ = Describe("VirtualMachine", func() {
+				sanityExecute(vm)
+
+				actions := virtFakeClient.Actions()
+				hasPatch := false
+				for _, action := range actions {
+					if action.GetVerb() == "patch" && action.GetResource().Resource == "virtualmachineinstances" {
+						hasPatch = true
+						break
+					}
+				}
+				Expect(hasPatch).To(BeFalse())
```
nit: maybe wrap this logic in a function and avoid all the repetitions

dasionov replied:
Thanks! fixed.

alicefr commented:
@dasionov question, is vm.Spec.Template.Spec.TerminationGracePeriodSeconds a live-updatable field?

dasionov commented:
> @dasionov question, is vm.Spec.Template.Spec.TerminationGracePeriodSeconds a live-updatable field?

by a live-updatable field you mean that this field will only be updated when the clusterConfig has liveUpdateRoloutStrategy?

cause i had a note in the description about how it wont make sense to me to configure it as live-update, but more of a always-update thing cause if we might have to wait for a reboot for it to affect the vmi, it will be pointless cause we aim to affect the close deletion of the virt-laucnher

wdyt?

dasionov commented:
/test pull-kubevirt-unit-test

0xFelix commented on pkg/virt-controller/watch/vm/vm.go:
```diff
@@ -3057,6 +3058,10 @@ func (c *Controller) sync(vm *virtv1.VirtualMachine, vmi *virtv1.VirtualMachineI
 	}
 
+	if err := c.handleTerminationGracePeriodRequest(vmCopy, vmi); err != nil {
+		return vm, vmi, common.NewSyncError(fmt.Errorf("error encountered while handling termination grace preiod request: %v", err), terminationGracePeriodErrorReason), nil
```
```suggestion
		return vm, vmi, common.NewSyncError(fmt.Errorf("error encountered while handling termination grace period request: %v", err), terminationGracePeriodErrorReason), nil
```

dasionov replied:
done

0xFelix commented on pkg/virt-controller/watch/vm/vm.go:
```diff
@@ -3262,3 +3267,28 @@ func (c *Controller) handleMemoryHotplugRequest(vm *virtv1.VirtualMachine, vmi *
+	desiredGracePeriod := vm.Spec.Template.Spec.TerminationGracePeriodSeconds
+	currentGracePeriod := vmi.Spec.TerminationGracePeriodSeconds
+
+	if desiredGracePeriod == nil || (currentGracePeriod != nil && *desiredGracePeriod == *currentGracePeriod) {
```
Should this result in the termination grace period being a live updatable field? What if `currentGracePeriod != nil` but `desiredGracePeriod` was intentionally set to `nil`?

dasionov replied:
IMO If that is what the user wants, then we shell remove the field in the VMI as well. the controller should be able to tolerate that, during migrations it shell use a default grace period value of 30 seconds, and in the delete options we will use `gracePeriod := pointer.P(int64(30))` unless `vmi.Spec.TerminationGracePeriodSeconds != nil`

0xFelix commented on pkg/virt-controller/watch/vm/vm_test.go:
```diff
@@ -241,6 +241,17 @@ var _ = Describe("VirtualMachine", func() {
+			hasPatch := false
+			for _, action := range actions {
+				if action.GetVerb() == "patch" && action.GetResource().Resource == "virtualmachineinstances" {
+					hasPatch = true
```
nit: Could be replace with `Fail("found patch when no patch was expected")`?

0xFelix replied:
Instead of the loop it could also be a reactor that fails when it is called.

0xFelix commented on pkg/virt-controller/watch/vm/vm_test.go:
```diff
@@ -6127,6 +6138,80 @@ var _ = Describe("VirtualMachine", func() {
 		})
 
+		Context("TerminationGracePeriodSeconds handling", func() {
+
```
nit: empty line

alicefr commented:
> > @dasionov question, is vm.Spec.Template.Spec.TerminationGracePeriodSeconds a live-updatable field?
> 
> by a live-updatable field you mean that this field will only be updated when the clusterConfig has liveUpdateRoloutStrategy?
> 
> cause i had a note in the description about how it wont make sense to me to configure it as live-update, but more of a always-update thing cause if we might have to wait for a reboot for it to affect the vmi, it will be pointless cause we aim to affect the close deletion of the virt-laucnher
> 
> wdyt?

I'm asking because how live-updatable fileds works. We usually check the old vm spec with the new one and we validate if the changed field can be live-updated. Now here, you are allowing changing a field which wasn't considered live-updatable. So, I'm wondering why you don't get the RestartRequired condition. I might missing something, but my question originates from this doubt

dasionov commented:
> > > @dasionov question, is vm.Spec.Template.Spec.TerminationGracePeriodSeconds a live-updatable field?
> > 
> > 
> > by a live-updatable field you mean that this field will only be updated when the clusterConfig has liveUpdateRoloutStrategy?
> > cause i had a note in the description about how it wont make sense to me to configure it as live-update, but more of a always-update thing cause if we might have to wait for a reboot for it to affect the vmi, it will be pointless cause we aim to affect the close deletion of the virt-laucnher
> > wdyt?
> 
> I'm asking because how live-updatable fields work. We usually check the old vm spec with the new one and we validate if the changed field can be live-updated. Now here, you are allowing changing a field which wasn't considered live-updatable. So, I'm wondering why you don't get the RestartRequired condition. I might miss something, but my question originates from this doubt

Fair question! I hope my answer addresses your concerns:

1. In my current implementation, I handle the grace period termination change request **after** validating and verifying the `RestartRequired` conditions. As a result, the condition is not being affected.
2. The field is not currently declared as `live-updatable`, but in my opinion, it should be—unless there's a valid reason for it to take effect only after a restart.

dasionov commented:
/test pull-kubevirt-goveralls

dasionov commented:
/test pull-kubevirt-e2e-k8s-1.32-sig-compute

0xFelix commented on pkg/virt-controller/watch/vm/vm_test.go:
```diff
@@ -6127,6 +6132,76 @@ var _ = Describe("VirtualMachine", func() {
+				Expect(err).ToNot(HaveOccurred())
+				addVirtualMachine(vm)
+
+				virtFakeClient.PrependReactor("patch", "virtualmachineinstances", func(action testing.Action) (handled bool, ret runtime.Object, err error) {
```
Instead of asserting the patch, you could assert the outcome instead. Get the VMI from the fake client and expect the `terminationGracePeriodSeconds` to be what is in the updated VM.

dasionov replied:
i was under the impression we couldn't do that, but found out a trick we made to allow that, so thanks !

0xFelix commented on pkg/virt-controller/watch/vm/vm_test.go:
```diff
@@ -6127,6 +6132,76 @@ var _ = Describe("VirtualMachine", func() {
+
+				Expect(controller.vmiIndexer.Delete(vmi)).To(Succeed())
+
+				virtFakeClient.PrependReactor("patch", "virtualmachineinstances", noPatchReactor)
```
You could additionally assert that the value was not changed on the VMI.

dasionov replied:
thanks! good call.

0xFelix commented on pkg/virt-controller/watch/vm/vm_test.go:
```diff
@@ -6127,6 +6132,76 @@ var _ = Describe("VirtualMachine", func() {
+				Expect(err).ToNot(HaveOccurred())
+				addVirtualMachine(vm)
+
+				virtFakeClient.PrependReactor("patch", "virtualmachineinstances", noPatchReactor)
```
Same

dasionov replied:
fixed

0xFelix commented on pkg/virt-controller/watch/vm/vm_test.go:
```diff
@@ -5339,6 +5339,86 @@ var _ = Describe("VirtualMachine", func() {
+
+					updatedVMI, err := virtFakeClient.KubevirtV1().VirtualMachineInstances(vm.Namespace).Get(context.TODO(), vmi.Name, metav1.GetOptions{})
+					Expect(err).ToNot(HaveOccurred())
+					Expect(updatedVMI.Spec.TerminationGracePeriodSeconds).ToNot(BeNil())
+					Expect(*updatedVMI.Spec.TerminationGracePeriodSeconds).To(Equal(int64(VMGracePeriod)))
```
nit: You could use the `HaveValue` matcher

dasionov replied:
done

0xFelix commented on pkg/virt-controller/watch/vm/vm_test.go:
```diff
@@ -5339,6 +5339,86 @@ var _ = Describe("VirtualMachine", func() {
+				})
+
+				It("should update VMI grace period when it differs from VM", func() {
+					const VMGracePeriod = 60
```
```suggestion
					const vmGracePeriod = 60
```

nit

dasionov replied:
fixed

0xFelix commented on pkg/virt-controller/watch/vm/vm_test.go:
```diff
@@ -5339,6 +5339,86 @@ var _ = Describe("VirtualMachine", func() {
+					Expect(*updatedVMI.Spec.TerminationGracePeriodSeconds).To(Equal(int64(vmiInitialGracePeriod)))
+				})
+
+				It("should do nothing when VM has no grace period specified", func() {
```
Shouldn't it keep VM and VMI in sync?

0xFelix commented on pkg/virt-controller/watch/vmi/vmi.go:
```diff
@@ -1905,3 +1906,11 @@ func (c *Controller) createPod(key, namespace string, pod *k8sv1.Pod) (*k8sv1.Po
+
+func (c *Controller) getDeleteOptionsForPod(pod *k8sv1.Pod, vmi *virtv1.VirtualMachineInstance) v1.DeleteOptions {
+	if strings.HasPrefix(pod.Name, "virt-launcher") && vmi.Spec.TerminationGracePeriodSeconds != nil {
+		gracePeriod := pointer.P(*vmi.Spec.TerminationGracePeriodSeconds + 30)
+		return v1.DeleteOptions{GracePeriodSeconds: gracePeriod}
```
```suggestion
		gracePeriod := *vmi.Spec.TerminationGracePeriodSeconds + 30
		return v1.DeleteOptions{GracePeriodSeconds: &gracePeriod}
```

0xFelix replied:
or inline it and keep `pointer.P`

dasionov commented:
> Once addressed Felix's comment, the rest looks good to me /approve

/hold

I just found out that we have a `virtctl stop` command that can be used like this:
`virtctl stop myvm --force --grace-period=0`

This command applies the --grace-period value from the flag, forcefully terminating the VMI and its launcher.
With this command available, I don't see a strong reason to keep the changes in this PR.

When I initially started working on this, I was under the impression that clients using a high grace period didn't have a convenient way to quickly terminate a VM when needed. Clearly, I was mistaken.

WDYT, @alicefr, @0xFelix?

0xFelix commented:
If virtctl can do this today without code changes, then we should close this PR :)

dasionov commented:
> If virtctl can do this today without code changes, then we should close this PR :)

Currently, if someone wants to quickly terminate a VM and bypass the termination grace period, they must use the `virtctl stop` command.

The key question here is: does it make sense to introduce the changes in this PR to allow future deletion of a VM without stopping it first?

In my opinion, the answer is no.
Without native Kubernetes support for a `grace-period` delete option, enforcing this behavior would require manually patching the field to propagate it to the VMI, which then affects the launcher deletion.

Additionally, creating a dedicated `virtctl` command for this purpose doesn’t seem justified when the existing `stop` command already handles termination as expected.

Therefore, I am closing this PR until a proper use case for these changes is introduced.
/close
